runtime/freelist.c

```diff
@@ -1308,40 +1308,46 @@ static int bf_is_in_tree (large_free_block *b)
 /**************************************************************************/
 
 /* Add back a remnant into a small free list. The block must be small
-   and black and its tag must be abstract. */
+   and white (or a 0-size fragment).
+   The block may be left out of the list depending on the sweeper's state.
+   The free list size is updated accordingly.
+*/
 static void bf_insert_remnant_small (value v)
 {
   mlsize_t wosz = Wosize_val (v);
 
-  CAMLassert (Color_val (v) == Caml_black);
-  CAMLassert (Tag_val (v) == Abstract_tag);
-  CAMLassert (1 <= wosz && wosz <= BF_NUM_SMALL);
-  Next_small (v) = bf_small_fl[wosz].free;
-  bf_small_fl[wosz].free = v;
-  if (bf_small_fl[wosz].merge == &bf_small_fl[wosz].free){
-    bf_small_fl[wosz].merge = &Next_small (v);
+  CAMLassert (Color_val (v) == Caml_white);
+  CAMLassert (wosz <= BF_NUM_SMALL);
+  if (wosz != 0 && (char *) Hp_val (v) < (char *) caml_gc_sweep_hp){
```

Comment:
I think we need to also check whether `caml_gc_phase == Phase_sweep` here.

Reply:
That's right. Fixed.